src/extract/global_greedy_dag.rs

```diff
@@ -0,0 +1,193 @@
+#[derive(Default)]
+pub struct TermDag {
+    nodes: Vec<Term>,
+    info: Vec<TermInfo>,
+    hash_cons: HashMap<Term, TermId>,
+}
+
+impl TermDag {
+    // Makes a new term using a node and children terms
+    // Correctly computes total_cost with sharing
+    // If this term contains itself, returns None
+    // If this term costs more than target, returns None
+    pub fn make(
+        &mut self,
+        node_id: NodeId,
+        node: &Node,
+        children: Vec<TermId>,
+        target: NotNan<f64>,
+    ) -> Option<TermId> {
+        let term = Term {
+            op: node.op.clone(),
+            children: children.clone(),
+        };
+
+        if let Some(id) = self.hash_cons.get(&term) {
+            return Some(*id);
+        }
+
+        let node_cost = node.cost;
+
+        if children.is_empty() {
+            let next_id = self.nodes.len();
+            self.nodes.push(term.clone());
+            self.info.push(TermInfo {
+                node: node_id,
+                eclass: node.eclass.clone(),
+                node_cost,
+                total_cost: node_cost,
+                reachable: [node.eclass.clone()].into_iter().collect(),
+                size: 1,
+            });
+            self.hash_cons.insert(term, next_id);
+            Some(next_id)
+        } else {
+            // check if children contains this node
+            for child in &children {
+                if self.info[*child].reachable.contains(&node.eclass) {
+                    return None;
+                }
+            }
+
+            let biggest_child = (0..children.len())
+                .max_by_key(|i| self.info[children[*i]].size)
+                .unwrap();
+
+            let mut cost = node_cost + self.total_cost(children[biggest_child]);
+            let mut reachable = Box::new(self.info[children[biggest_child]].reachable.clone());
+            let next_id = self.nodes.len();
+
+            for child in children.iter() {
+                if cost > target {
+                    return None;
+                }
+                let child_cost = self.get_cost(&mut reachable, *child);
+                cost += child_cost;
+            }
+
+            if cost > target {
+                return None;
+            }
+
+            *reachable = reachable.insert(node.eclass.clone());
+
+            self.info.push(TermInfo {
+                node: node_id,
+                node_cost,
+                eclass: node.eclass.clone(),
+                total_cost: cost,
+                reachable: *reachable,
+                size: 1 + children.iter().map(|c| self.info[*c].size).sum::<usize>(),
+            });
+            self.nodes.push(term.clone());
+            self.hash_cons.insert(term, next_id);
+            Some(next_id)
+        }
+    }
+
+    // Return a new term, like this one but making use of shared terms.
+    // Also return the cost of the new nodes.
+    fn get_cost(&self, shared: &mut Box<Reachable>, id: TermId) -> NotNan<f64> {
```

Comment:
nit: let's call this `Cost` (I think that may be its name elsewhere in this repo)